 <h1 style="text-align: center; color: RED">AB TESTING</h1>

In [6]:
import itertools
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.stats.api as sms
from scipy.stats import ttest_1samp, shapiro, levene, ttest_ind, mannwhitneyu, \
    pearsonr, spearmanr, kendalltau, f_oneway, kruskal
from statsmodels.stats.proportion import proportions_ztest

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 10)
pd.set_option('display.float_format', lambda x: '%.5f' % x)

 <h2 style="text-align: center; color: RED">HİPOTEZ TESTLERİ</h2>

Hipotez; bir inanışı, bir savı test etmek için kullanılan istatistiksel yöntemlerdir. Bizim odaklanacak olduğumuz konu ise:
Grup karşılaştırmalarında temel amaç olası farklılıkların şans eser ortaya çıkıp çıkmadığını göstermeye çalışmaktır.


Grup karşılaştırılmalarında temel amaç olası farklılıkların şans eser ortaya çıkıp çıkmadığını göstermeye çalışmaktır. Örneğin şöyle bir savımız olsun:
Mobil uygulamada yapılan arayüz değişikliği sonrasında kullanıcıların uygulamada geçirdikleri günlük ortalama süre arttı mı?

Yukarıdaki sav için ilk arayüz A grubundaki örneklemlere denetilsin, B grubundaki örneklemelre ise yeni arayüz denetilsin.Çıkan sonuÇ A grubu için 55 dk iken B grubu için 58 dk olsun. Şimdi şu soruyu sorabiliriz iki arayüz arasında gerçekten bir fark var mı yoksa oluşan bu fark rastgele mi oluştu? İşte tam bu noktada AB test devreye girmektedir.

<span style="color: blue; font-weight: bold;">1-AB TESTİ(BAĞIMSIZ İKİ ÖRNEKLEM T TESTİ)</span>

İki grup ortalaması arasında karşılaştırma yapılmak istenildiğinde kullanılır. Bu test, grupların ortalamalarının birbirinden farklı olup olmadığını belirlemeye yardımcı olur.

![Test](datasets/3.png)

H0:Yokluk hipotezidir. Sınayacak olduğumuz durum H0 hipotezindedir.Yani iki grup ortalaması arasında fark yoktur.
Biz bu H0 hipotezini sınarız.H0 hipotezini reddetme veya reddedememeye göre iki grup ortalaması arasında fark olup olmadığını değerlendiririz.

Kullanılacak formüller:
![Test](datasets/4.png)
Fakat biz bu formüllerle uğrşamayıp p < 0.05 üzerinden ilerleyecez.

️**Bağımsız İki Örneklem T Testinin Varsayımları:**

*Normallik:* Her iki gruptaki veri setinin normal dağıldığı varsayılır. shapiro testi ile yapılır.

*Varyans Homojenliği:* İki grubun varyanslarının benzer olduğu kabul edilir. Eğer varyanslar eşit değilse, Welch'in t testi kullanılır.


* Adım 1: Hipotez Kurma
Bu test için iki tür hipotez kurulur:
Null Hipotezi (H₀): İki grubun ortalamaları arasında fark yoktur.
H0:μ1=μ2
Alternatif Hipotez (H₁): İki grubun ortalamaları arasında fark vardır.
H1:μ1≠μ2

* Adım 2: Testin Uygulanacağı Varsayımları Kontrol Etme
Bağımsızlık: İki grup bağımsız olmalıdır (örneğin, bir grubun diğer grubun elemanları ile ilişkisi olmamalıdır).
Normal Dağılım: Veriler, her iki grupta da yaklaşık olarak normal dağılmalıdır.
Varyansların Eşitliği (Opsiyonel): Eğer varyansların eşit olduğu varsayımı sağlanıyorsa, eşit varyanslı test kullanılır. Aksi halde, eşit olmayan varyanslı test kullanılır.

* Adım 3: Test İstatistiğini Hesaplama

* Adım 4:Test istatistiği ile p-değeri hesaplanır. Bu değeri kullanarak null hipotez reddedilip reddedilmeyeceğine karar verilir.Eğer p < 0.05 ise null hipotez reddedilir, yani gruplar arasında anlamlı bir fark vardır.Eğer p ≥ 0.05 ise null hipotez reddedilemez, yani gruplar arasında anlamlı bir fark yoktur.

* Adım 5: Sonuçları Yorumlama
p-değeri ve t-istatistiği kullanılarak sonuca varılır ve hipotez testinin sonucuna göre karar verilir.



<span style="color: blue; font-weight: bold;">2-AB TESTİ(BAĞIMSIZ İKİ ÖRNEKLEM T TESTİ) UYGULAMA-1</span>

**Uygulamalarda izlenecek adımlar:**
* 1. Hipotezleri Kur
* 2. Varsayım Kontrolü
*    - 1. Normallik Varsayımı:dağılımlar normal
*    - 2. Varyans Homojenliği:varyansaların benzer olması
* 3. Hipotezin Uygulanması
*    - 1. Varsayımlar sağlanıyorsa bağımsız iki örneklem t testi (parametrik test)
*    - 2. Varsayımlar sağlanmıyorsa mannwhitneyu testi (non-parametrik test)
* 4. p-value değerine göre sonuçları yorumla
     
* Not:
- Normallik sağlanmıyorsa direk 2 numara. Varyans homojenliği sağlanmıyorsa 1 numaraya arguman girilir.

- Normallik incelemesi öncesi aykırı değer incelemesi ve düzeltmesi yapmak faydalı olabilir.

Yukarıdaki işlemleri ab testi eğer iki grup ortalamasına yönelikse bu adımları uygulayabiliriz.

In [31]:
############################
# Uygulama 1: Sigara İçenler ile İçmeyenlerin Hesap Ortalamaları Arasında İstatistiksel Olarak Fark Var Mı?
############################

In [37]:
df = sns.load_dataset('tips')

In [39]:
df.head()

total_bill     tip     sex smoker  day    time  size
0    16.99000 1.01000  Female     No  Sun  Dinner     2
1    10.34000 1.66000    Male     No  Sun  Dinner     3
2    21.01000 3.50000    Male     No  Sun  Dinner     3
3    23.68000 3.31000    Male     No  Sun  Dinner     2
4    24.59000 3.61000  Female     No  Sun  Dinner     4

In [41]:
df.groupby('smoker').agg({'total_bill':'mean'})

C:\Users\ASLI\AppData\Local\Temp\ipykernel_16452\1864663744.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('smoker').agg({'total_bill':'mean'})


total_bill
smoker            
Yes       20.75634
No        19.18828

* Yukarıdaki ortalamalara bakıldıgında sigara içenlerin ödediği hesap ile içmeyenleirn ödedği hesap arasında fark var.Peki gerçekten istatistiksel bir fark var mı?

**Adım1:** Hipotez kur

H0=m1 = m2 (sigara içen ile içmeyenlerin ödediği hesaplar arasında fark yoktur.)

H1= m1!=m2

**Adım2:** Varsayım Kontrolü

* Normallik Varsayımı(bir değişkenin dağılımının standart normal dağılıma benzer olup olmadığının hipotez testidir.)

  - H0: Normal dağılım varsayımı sağlanmaktadır.

  - H1:..sağlanmamaktadır.
* Varyans Homojenliği


In [57]:
#Normallik varsayımı
test_stat, pvalue = shapiro(df.loc[df['smoker']=='Yes','total_bill'])
print('Test Stat = %.4f, p-value=%.4f' % (test_stat,pvalue))

Test Stat = 0.9367, p-value=0.0002


* p < 0.05 ise ise H0 reddedilir. Değilse reddedilemez.
  
0.0002 < 0.05 olduğu için H0 reddidilir. Yani normal dağılım varsayımı sağlanmamaktadır.

In [60]:
test_stat, pvalue = shapiro(df.loc[df['smoker']=='No','total_bill'])
print('Test Stat = %.4f, p-value=%.4f' % (test_stat,pvalue))

Test Stat = 0.9045, p-value=0.0000


0.0000 < 0.05 olduğu için H0 reddidilir. Yani normal dağılım varsayımı sağlanmamaktadır.

**Normallik varsayımı sağlanmadığı için non-paramertrik testi kullanacaz.** 

Varsayalımki normallik varsayımı sağlandı, sağlandıktan sonraki aşama ne olurdu?

In [73]:
#Varyans homojenliği
#H0: Varyanslar homojendir
#H1: Varyanslar homojen değildir.
test_stat, pvalue = levene(df.loc[df['smoker']=='Yes', 'total_bill'],
                           df.loc[df['smoker']=='No', 'total_bill'])
print('Test Stat = %.4f, p-value=%.4f' % (test_stat,pvalue))

Test Stat = 4.0537, p-value=0.0452


0.0452 < 0.05 olduğu için H0 reddidilir. Yani varyans homojenliği sağlanmamaktadır.

İki varyansta sağlanmamaktır. Bu durumda non-parametrik testi kullanırız

**Adım:** Hipotezin Uygulanması

1-Varsayımlar sağlanıyorsa bağımsız iki örneklem t testi (parametrik test)

2-Varsayımlar sağlanmıyorsa mannwhitneyu testi (non-parametrik test)

In [84]:
#Varsayımlar sağlanmış gibi davranıp parametrik testi yapalım
#T testi:
# normallik varsayı sağlanıyorsa, 
#her iki varsayım sağlanıyorsa, 
#normallik varsayımı sağlanıyor varyans homojenliği sağlanmıyorsa da kullanılır fakat equa_var=True parametersi False yapılmalıdır.
test_stat, pvalue = ttest_ind(df.loc[df['smoker']=='Yes', 'total_bill'],
                              df.loc[df['smoker']=='No', 'total_bill'],
                              equal_var=True)
print('Test Stat = %.4f, p-value=%.4f' % (test_stat,pvalue))

Test Stat = 1.3384, p-value=0.1820


p < 0.05 oldugunda H0 reddidilir.
0.1820 < 0.05 olmadığından dolayı H0 reddedilemez. Yani ortalamaları arasında bir fark yoktur deriz.Oluşan bir fark varsa bu şans eseri oluşmuştur diyebiliriz.

In [90]:
#2 varyans sağlanmadığı için asıl yapmamız gereken test non-parametrik testtir.

test_stat, pvalue = mannwhitneyu(df.loc[df["smoker"] == "Yes", "total_bill"],
                                 df.loc[df["smoker"] == "No", "total_bill"])

print('Test Stat = %.4f, p-value = %.4f' % (test_stat, pvalue))

Test Stat = 7531.5000, p-value = 0.3413


p < 0.05 oldugunda H0 reddidilir.  0.3413 < 0.05 olmadığından dolayı H0 reddedilemez. Yani ortalamaları arasında bir fark yoktur deriz.Oluşan bir fark varsa bu şans eseri oluşmuştur diyebiliriz.

<span style="color: blue; font-weight: bold;">2-AB TESTİ(BAĞIMSIZ İKİ ÖRNEKLEM T TESTİ) UYGULAMA-2</span>

In [4]:
############################
# Uygulama 2: Kadın ve Erkek Yolcuların Yaş Ortalamaları Arasında İstatistiksel Olarak Fark Var Mı?
############################

In [93]:
df = sns.load_dataset('titanic')

In [95]:
df.head()

survived  pclass     sex      age  sibsp  parch     fare embarked  class  \
0         0       3    male 22.00000      1      0  7.25000        S  Third   
1         1       1  female 38.00000      1      0 71.28330        C  First   
2         1       3  female 26.00000      0      0  7.92500        S  Third   
3         1       1  female 35.00000      1      0 53.10000        S  First   
4         0       3    male 35.00000      0      0  8.05000        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [97]:
df.groupby('sex')['age'].mean()#ortalamalar arasında bir fark var gibi görünüyor.

sex
female   27.91571
male     30.72664
Name: age, dtype: float64

**Uygulamalarda izlenecek adımlar:**
* 1. Hipotezleri Kur
* 2. Varsayım Kontrolü
*    - 1. Normallik Varsayımı:dağılımlar normal
*    - 2. Varyans Homojenliği:varyansaların benzer olması
* 3. Hipotezin Uygulanması
*    - 1. Varsayımlar sağlanıyorsa bağımsız iki örneklem t testi (parametrik test)
*    - 2. Varsayımlar sağlanmıyorsa mannwhitneyu testi (non-parametrik test)
* 4. p-value değerine göre sonuçları yorumla
     
* Not:
- Normallik sağlanmıyorsa direk 2 numara. Varyans homojenliği sağlanmıyorsa 1 numaraya arguman girilir.

- Normallik incelemesi öncesi aykırı değer incelemesi ve düzeltmesi yapmak faydalı olabilir.

Yukarıdaki işlemleri ab testi eğer iki grup ortalamasına yönelikse bu adımları uygulayabiliriz.

In [17]:
#1)HİPOTEZLERİ KUR
#H0:m1=2 (Kadın ve erkeklerin yaş ortalamaları arasında fark yoktur.)
#H1:m1!=m2 (Fark vardır.)

In [99]:
#2)VARSAYIMLARI İNCELE
####Normallik Varsayımı####
#- H0: Normal dağılım varsayımı sağlanmaktadır.
#- H1:..sağlanmamaktadır.

test_stat, pvalue = shapiro(df.loc[df['sex']=='female', 'age'].dropna())
print('Test_stat = %.4f, p-value=%.4f' %(test_stat,pvalue))

Test_stat = 0.9848, p-value=0.0071


* p < 0.05 ise ise H0 reddedilir. Değilse reddedilemez.

  0.0071 < 0.05 oldugundan h0 reddedilir. Yani normal dağılım varsayımı sağlanmamaktadır.

In [101]:
test_stat, pvalue = shapiro(df.loc[df['sex']=='male','age'].dropna())
print('Test_stat = %.4f, p-value=%.4f' %(test_stat,pvalue))

Test_stat = 0.9747, p-value=0.0000


  0.0000 < 0.05 oldugundan h0 reddedilir. Yani normal dağılım varsayımı sağlanmamaktadır.



In [103]:
#####Varyans Homojenliği#####
#H0: Varyanslar homojendir
#H1: Varyanslar homojen değildir.

test_stat, pvalue = levene(df.loc[df['sex']=='female', 'age'].dropna(),
                           df.loc[df['sex']=='male', 'age'].dropna())
print('Test Stat = %.4f, p-value=%.4f' % (test_stat,pvalue))

Test Stat = 0.0013, p-value=0.9712


0.9712 < 0.05 olamdığı için Ho reddeilemez yani varyanslar homojensir. 

Fakat normal dağılım sağlanmadığı için non parametrik test yapılır.

In [44]:
#2)HİPOTEZİN UYGULANMASI(mannwhitneyu testi)


In [107]:
test_stat, pvalue = mannwhitneyu(df.loc[df["sex"] == "female", "age"].dropna(),
                                 df.loc[df["sex"] == "male", "age"].dropna())

print('Test Stat = %.4f, p-value = %.4f' % (test_stat, pvalue))

Test Stat = 53212.5000, p-value = 0.0261


* p < 0.05 ise ise H0 reddedilir. Değilse reddedilemez.
  
0.0261 < 0.05 oldugu için H0 reddedilir yani yaş ortalamaları arasında istatistiksel bir fark vardır.

<span style="color: blue; font-weight: bold;">2-AB TESTİ(BAĞIMSIZ İKİ ÖRNEKLEM T TESTİ) UYGULAMA-3</span>

In [57]:
############################
# Uygulama 3: Diyabet Hastası Olan ve Olmayanların Yaşları Arasında İstatistiksel Olarak Fark Var Mı?
############################

In [61]:
df = pd.read_csv("C:\\Users\\ASLI\\Desktop\\Data Scientist Bootcamp\\4.Measurement Problems\\datasets\\diabetes.csv")

In [63]:
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin      BMI  \
0            6      148             72             35        0 33.60000   
1            1       85             66             29        0 26.60000   
2            8      183             64              0        0 23.30000   
3            1       89             66             23       94 28.10000   
4            0      137             40             35      168 43.10000   

   DiabetesPedigreeFunction  Age  Outcome  
0                   0.62700   50        1  
1                   0.35100   31        0  
2                   0.67200   32        1  
3                   0.16700   21        0  
4                   2.28800   33        1

In [65]:
df.groupby('Outcome')['Age'].mean()#ortalamar arasında bir fark var görünüyor peki bu fark gercekten var mı?

Outcome
0   31.19000
1   37.06716
Name: Age, dtype: float64

In [67]:
#1)HİPOTEZLERİ KUR
#H0:m1=2 (Diyabet olanlar ile olmayanların yaş ortalamaları arasında fark yoktur.)
#H1:m1!=m2 (Fark vardır.)

In [71]:
#2)VARSAYIMLARI İNCELE
####Normallik Varsayımı####
#- H0: Normal dağılım varsayımı sağlanmaktadır.
#- H1:..sağlanmamaktadır.
test_stat, pvalue = shapiro(df.loc[df['Outcome']==0,'Age'].dropna())
print('Test_stat= %.4f, p-value= %.4f' % (test_stat, pvalue))

Test_stat= 0.8012, p-value= 0.0000


p < 0.05 H0 ise H0 reddedilir.

0.0000< 0.05 oldugundan H0 reddedilir. Yani varsayım sağlanmamaktadır.

In [74]:
test_stat, pvalue = shapiro(df.loc[df['Outcome']==1,'Age'].dropna())
print('Test_stat= %.4f, p-value= %.4f' % (test_stat, pvalue))

Test_stat= 0.9546, p-value= 0.0000


0.0000< 0.05 oldugundan H0 reddedilir. Yani varsayım sağlanmamaktadır.

In [79]:
#####Varyans Homojenliği#####
#H0: Varyanslar homojendir
#H1: Varyanslar homojen değildir.

test_stat, pvalue = levene(df.loc[df['Outcome']==0,'Age'].dropna(),
                           df.loc[df['Outcome']==1,'Age'].dropna())
print('Test_stat= %.4f, p-value= %.4f' % (test_stat, pvalue))

Test_stat= 2.2252, p-value= 0.1362


0.1362 < 0.05 olmadığından H0 reddedilemez yani varyan homojenliği sağlanmaktadır. Normallik varsayımı sağlanmadığı için non parametrik test olan mannwhitneyu testi yapılacaktır.

In [84]:
#3)HİPOTEZİN UYGULANMASI(mannwhitneyu testi)
test_stat, pvalue = mannwhitneyu(df.loc[df['Outcome']==0,'Age'].dropna(),
                                 df.loc[df['Outcome']==1,'Age'].dropna())
print('Test_stat= %.4f, p-value= %.4f' % (test_stat, pvalue))

Test_stat= 41950.0000, p-value= 0.0000


0.000 < 0.05 oldugundan H0 reddedilir. Yani diyabet olanlar ile olmayanların yaş ortalamaları arasında istatistiksel bir fark vardr. 

<span style="color: blue; font-weight: bold;">2-AB TESTİ(BAĞIMSIZ İKİ ÖRNEKLEM T TESTİ) UYGULAMA-4</span>

In [113]:
############################
# Uygulama 4: Kursun Büyük Çoğunluğunu İzleyenler İle İzlemeyenlerin Puanları Birbirinden Farklı mı?
############################

In [115]:
df = pd.read_csv("C:\\Users\\ASLI\\Desktop\\Data Scientist Bootcamp\\4.Measurement Problems\\datasets\\course_reviews.csv")

In [117]:
df.head()

Rating            Timestamp             Enrolled  Progress  \
0 5.00000  2021-02-05 07:45:55  2021-01-25 15:12:08   5.00000   
1 5.00000  2021-02-04 21:05:32  2021-02-04 20:43:40   1.00000   
2 4.50000  2021-02-04 20:34:03  2019-07-04 23:23:27   1.00000   
3 5.00000  2021-02-04 16:56:28  2021-02-04 14:41:29  10.00000   
4 4.00000  2021-02-04 15:00:24  2020-10-13 03:10:07  10.00000   

   Questions Asked  Questions Answered  
0          0.00000             0.00000  
1          0.00000             0.00000  
2          0.00000             0.00000  
3          0.00000             0.00000  
4          0.00000             0.00000

In [123]:
df[(df['Progress']>75)]['Rating'].mean()#Kursun %75'ten fazla izleyenlerin verdiği puanların ortalaması

4.860491071428571

In [128]:
df[(df['Progress']<25)]['Rating'].mean()#Kursu %10'dan az izleyenlerin verdiği puanların ortalaması

4.7225029148853475

Ortalamalara baktığımızda sanki kursu daha çok izleyenler kursu daha iyi tanıyıp puanlama yapmış gibi.

In [130]:
##1)HİPOTEZLERİ KUR
#H0:m1=2 (Kursun çoğunluğunu izleyneler ile izlemeyenlerin verdiği puanlar arasında fark yoktur.)
#H1:m1!=m2 (Fark vardır.)

In [132]:
#2)VARSAYIMLARI İNCELE
####Normallik Varsayımı####
#- H0: Normal dağılım varsayımı sağlanmaktadır.
#- H1:..sağlanmamaktadır.

test_stat, pvalue = shapiro(df.loc[df['Progress'] < 25,'Rating'].dropna())
print('Test_stat= %.4f, p-value= %.4f' % (test_stat, pvalue))

Test_stat= 0.5710, p-value= 0.0000


In [134]:
test_stat, pvalue = shapiro(df.loc[df['Progress'] >75,'Rating'].dropna())
print('Test_stat= %.4f, p-value= %.4f' % (test_stat, pvalue))

Test_stat= 0.3160, p-value= 0.0000


p değerleri 0.05 den küçük oldugu için H0 reddedilir. Yani normal dağılım sağlanmamaktadır. Normal dağılım sağlanmadığı için varyans homojenliğine bakmaya gerek yoktur. Non-parametrik test yapılacaktır.

In [137]:
#3)HİPOTEZİN UYGULANMASI(mannwhitneyu testi)
test_stat, pvalue = mannwhitneyu(df.loc[df['Progress'] < 25,'Rating'].dropna(),
                                 df.loc[df['Progress'] >75,'Rating'].dropna())
print('Test_stat= %.4f, p-value= %.4f' % (test_stat, pvalue))

Test_stat= 491222.5000, p-value= 0.0000


0.0000 < 0.05 olduğundan H0 reddedilir. Yani kursun çoğunluğunu izleyenler ile izlemeyenlerin verdiği puanlar arasında istatistiksel fark vardır.